### PR TITLE
[GNA] Improve performance for PWL algorithm

### PR DESCRIPTION
**Pull request by dmitriikhurtin**
### Details:
 - Binary search is used to search for segments.

### Tickets:
 - CVS-70363

### CODE REVIEW

File: inference-engine/src/gna_plugin/runtime/pwl.cpp [hunk 1]
```diff
@@ -401,98 +472,62 @@ std::vector<pwl_t> pwl_search(const DnnActivation& activation_type,
...
+
+            int n_segments_mid = n_segments_lower + (n_segments_upper - n_segments_lower) / 2;
+            std::pair<double, int> previous_good_result(err_pct, n_segments_mid);
+            while (std::abs(n_segments_lower - n_segments_upper) > 1) {
+                if (!pivot_search(pwl, activation_type, n_segments_mid, l_bound, u_bound, threshold, err, max_iteration_number)) {
+                    n_segments_upper = n_segments_mid;
+                    err_pct = std::numeric_limits<double>::max();
+                } else {
+                    err_pct = calculate_error_pct(activation_type, l_bound, u_bound, err, samples);
+                    if (allowed_err_pct == err_pct) {
```
Comment (nkogteva) [inline inference-engine/src/gna_plugin/runtime/pwl.cpp:497]: I think the comparison of two double variables to equality has no sense. It is better to compare with some epsilon values.

**Review by nkogteva — COMMENTED** [review]
I think we need to perform some measurements for this branch. At least load time for some topologies before/after.

File: inference-engine/src/gna_plugin/runtime/pwl.cpp (hunk 1 above)
Comment (dmitriikhurtin) [reply inference-engine/src/gna_plugin/runtime/pwl.cpp]: done

File: inference-engine/src/gna_plugin/runtime/pwl.cpp [hunk 2]
```diff
@@ -168,28 +178,79 @@ double pivot_search(std::vector<pwl_t>& result,
...
+                  double& epsilon_final,
+                  const int max_iteration_number) {
+    bool negative = false;
+    double(*f)(const double) = nullptr;
+    double(*first_deriv_f)(const double) = nullptr;
+    switch (activation_type) {
+        case kActSigmoid:
+            if (alpha_N == 0) negative = true;  // make left half convex
+            f = sigmoid;
+            first_deriv_f = first_deriv_sigmoid;
```
[the remark this answers is not in the document]
Comment (dmitriikhurtin) [reply inference-engine/src/gna_plugin/runtime/pwl.cpp:201]: I'm planning remove all the switches with activation_type when we move to pwl op.

**Review by elilobanova — APPROVED** [review]

File: inference-engine/src/gna_plugin/runtime/pwl.cpp [hunk 3]
```diff
@@ -401,98 +473,54 @@ std::vector<pwl_t> pwl_search(const DnnActivation& activation_type,
...
-                    default:
-                        break;
                 }
-                err_pct = calculate_error_pct(activation_type, l_bound, u_bound, err, samples);
-            }
-
-            if (n_segments >= PWL_MAX_ITERATIONS) {
-                THROW_GNA_EXCEPTION << "Failed to converge in pwl_search!";
+            } while (n_segments_lower <= n_segments_upper);
+
```
[the remark this answers is not in the document]
Comment (dmitriikhurtin) [reply inference-engine/src/gna_plugin/runtime/pwl.cpp:511]: I think this information will be useful.

**Review by nkogteva — APPROVED** [review]

File: inference-engine/src/gna_plugin/runtime/pwl.h [hunk 4]
```diff
@@ -23,7 +23,7 @@
 #define PWL_MAX_ERR_PERCENT 1.0f
 #define PWL_MAX_ITERATIONS_DEFAULT 2000
 #define PWL_MAX_ITERATIONS_LOG 5000
-#define PWL_MAX_NUM_SEGMENTS 128
+#define PWL_MAX_NUM_SEGMENTS (128-2)
```
Comment (mryzhov) [inline inference-engine/src/gna_plugin/runtime/pwl.h:26]: what is the magic number 2? Looks like a debug change

Comment (dmitriikhurtin) [reply inference-engine/src/gna_plugin/runtime/pwl.h:26]: In some cases, two segments are added to the beginning and end when creating a PWL on the GNU side (see create_multi segment_gna_pwl() in inference-engine/src/gna_plugin/backend/make_pwl.cpp ).

**Comment by akladiev** [comment]
This PR will be closed in 2 weeks in case of no activity.

**Pull request closed without merging**